sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs

```diff
@@ -0,0 +1,255 @@
+            Queue<string> years,
+            Queue<string> segments,
+            Segment currentSegment,
+            DateTimeOffset lastConsumable,
+            DateTimeOffset? startTime,
+            DateTimeOffset? endTime)
+        {
+            _containerClient = containerClient;
+            _segmentFactory = segmentFactory;
+            _years = years;
+            _segments = segments;
+            _currentSegment = currentSegment;
+            _lastConsumable = lastConsumable;
+            _startTime = startTime;
+            _endTime = endTime;
+            _empty = false;
+        }
+
+        /// <summary>
+        /// Constructor for mocking, and for creating a Change Feed with no Events.
+        /// </summary>
+        public ChangeFeed() { }
+
+        // The last segment may still be adding chunks.
+        public async Task<Page<BlobChangeFeedEvent>> GetPage(
+            bool async,
+            int pageSize = Constants.ChangeFeed.DefaultPageSize,
+            CancellationToken cancellationToken = default)
+        {
+            if (!HasNext())
+            {
+                throw new InvalidOperationException("Change feed doesn't have any more events");
+            }
+
+            if (_currentSegment.DateTime > _endTime)
+            {
+                return BlobChangeFeedEventPage.Empty();
+            }
+
+            if (!_currentSegment.Finalized)
+            {
+                return BlobChangeFeedEventPage.Empty();
+            }
+
+            if (pageSize > Constants.ChangeFeed.DefaultPageSize)
+            {
+                pageSize = Constants.ChangeFeed.DefaultPageSize;
+            }
+
+            // Get next page
+            List<BlobChangeFeedEvent> blobChangeFeedEvents = new List<BlobChangeFeedEvent>();
+
+            int remainingEvents = pageSize;
+            while (blobChangeFeedEvents.Count < pageSize
+                && HasNext())
+            {
+                List<BlobChangeFeedEvent> newEvents = await _currentSegment.GetPage(
+                    async,
+                    remainingEvents,
+                    cancellationToken).ConfigureAwait(false);
+                blobChangeFeedEvents.AddRange(newEvents);
+                remainingEvents -= newEvents.Count;
+                await AdvanceSegmentIfNecessary(async).ConfigureAwait(false);
+            }
+
+            return new BlobChangeFeedEventPage(blobChangeFeedEvents, JsonSerializer.Serialize<ChangeFeedCursor>(GetCursor()));
+        }
+
+        public bool HasNext()
+        {
+            if (_empty)
+            {
+                return false;
+            }
+
+            // We have no more segments, years, and the current segment doesn't have hext.
+            if (_segments.Count == 0 && _years.Count == 0  && !_currentSegment.HasNext())
+            {
+                return false;
+            }
+
+            DateTimeOffset end = BlobChangeFeedExtensions.MinDateTime(_lastConsumable, _endTime);
```

Comment:
I think we should make the logic of `HasNext()` consistent with that in the `GetPage()`, i.e. ignore `_lastConsumable` and use `_currentSegment.Finalized`.
Or the iterator will end up returning empty change feed event page forever.

Reply:
Fixed.